File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
from collections import defaultdict
from heapq import heappop, heappush


counter = defaultdict(int)
max_heap, min_heap = [], []

def parse(command):
    c, num = command.split(" ")
    num = int(num)
    if c == 'I':
        counter[num] += 1
        heappush(max_heap, -num)
        heappush(min_heap, num) 
    else:
        # 최댓값 삭제
        if num == 1:
            while max_heap:
                removed = heappop(max_heap)
                # 존재하면 삭제 카운팅, 아니면 존재하는 것 끝까지 찾기
                if counter[-removed] > 0:
                    counter[-removed] -= 1
                    break
        # 최솟값 삭제
        else:
             while min_heap:
                removed = heappop(min_heap)
                # 존재하면 삭제 카운팅, 아니면 존재하는 것 끝까지 찾기
                if counter[removed] > 0:
                    counter[removed] -= 1
                    break
    #print(max_heap, min_heap, counter)
        
def solution(operations):
    '''
    Counter를 두어
    Insert 시에 개수 ++
    pop 시에 개수 --
    
    지우려고 하는 값의 개수가 0이 아닐 때 까지 pop 하기
    그러다가 heap이 empty이면 끝
    
    최댓값: 최대 힙에서 유효한 녀석 구하기
    최솟값: 최소 힙에서 유효한 녀석 구하기
    '''
    for o in operations:
        parse(o)
    
    max_result, min_result = 0,0
    while max_heap:
        popped = heappop(max_heap)
        if counter[-popped] > 0:
            max_result = -popped
            break
    while min_heap:
        popped = heappop(min_heap)
        if counter[popped] > 0:
            min_result = popped
            break
    return [max_result, min_result]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;
class Solution {
    /*
    공백 기준 앞이 D인지 아닌지
    */
    
    public static void interPrinter(String command, PriorityQueue<Integer> maxHeap, PriorityQueue<Integer> minHeap, Map<Integer,Integer> counter){
        String[] str = command.split(" ");
        if(str[0].equals("D")){
            if(str[1].equals("1")){
                while(!maxHeap.isEmpty()){
                    int value = maxHeap.poll();
                    int cnt = counter.get(value);
                    if(cnt > 0){
                        //System.out.println("최댓값 추출" + value + " -> " + (cnt-1));
                        counter.put(value, cnt - 1);
                        break;
                    }
                }
            }
            else{
                while(!minHeap.isEmpty()){
                    int value = minHeap.poll();
                    int cnt = counter.get(value);
                    if(cnt > 0){
                        //System.out.println("최솟값 추출" + value + " -> " + (cnt-1));
                        counter.put(value, cnt - 1);
                        break;
                    }
                }
            }
        }
        else{
            // push 하는 경우
            int num = Integer.parseInt(str[1]);
            maxHeap.offer(num);
            minHeap.offer(num);
            if(counter.containsKey(num)){
                counter.put(num, counter.get(num) + 1);
            }
            else{
                counter.put(num, 1);
            }
        }
    }
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minHeap = new PriorityQueue();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a,b) -> b-a);
        Map<Integer, Integer> counter = new HashMap<>();
        
        
        for(String s: operations){
            interPrinter(s, maxHeap, minHeap, counter);
        }
        //counter.forEach((k,v) -> System.out.println(k + " = " + v));
        int max = 0;
        int min = 0;

        // max 값 꺼내기
        while (!maxHeap.isEmpty()) {
            int value = maxHeap.peek();
            if (counter.getOrDefault(value, 0) > 0) {
                max = value;
                break;
            } else {
                maxHeap.poll(); // 쓰레기값 버리기
            }
        }

        // min 값 꺼내기
        while (!minHeap.isEmpty()) {
            int value = minHeap.peek();
            if (counter.getOrDefault(value, 0) > 0) {
                min = value;
                break;
            } else {
                minHeap.poll(); // 쓰레기값 버리기
            }
        }

        
        return new int[]{max, min};
    }
}